File: web/share/js/kvm/atx.js
/*****************************************************************************
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2024  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
*****************************************************************************/


"use strict";


import {tools, $} from "../tools.js";
import {wm} from "../wm.js";


export function Atx(__recorder) {
	var self = this;

	/************************************************************************/

	var __has_switch = null; // Or true/false
	var __state = null;

	var __init__ = function() {
		$("atx-power-led").title = "Power Led";
		$("atx-hdd-led").title = "Disk Activity Led";

		tools.storage.bindSimpleSwitch($("atx-ask-switch"), "atx.ask", true);

		tools.el.setOnClick($("atx-power-button"), () => __clickAtx("power"));
		tools.el.setOnClick($("atx-power-button-long"), () => __clickAtx("power_long"));
		tools.el.setOnClick($("atx-reset-button"), () => __clickAtx("reset"));
	};

	/************************************************************************/

	self.setState = function(state) {
		if (state) {
			if (!__state) {
				__state = {"leds": {}};
			}
			if (state.enabled !== undefined) {
				__state.enabled = state.enabled;
				tools.feature.setEnabled($("atx-dropdown"), (__state.enabled && !__has_switch));
			}
			if (__state.enabled !== undefined) {
				if (state.busy !== undefined) {
					__updateButtons(!state.busy);
					__state.busy = state.busy;
				}
				if (state.leds !== undefined) {
					__state.leds = state.leds;
				}
				if (state.busy !== undefined || state.leds !== undefined) {
					__updateLeds(__state.leds.power, __state.leds.hdd, __state.busy);
				}
			}
		} else {
			__state = null;
			__updateLeds(false, false, false);
			__updateButtons(false);
		}
	};

	self.setHasSwitch = function(has_switch) {
		__has_switch = has_switch;
		self.setState(__state);
	};

	var __updateLeds = function(power, hdd, busy) {
		$("atx-power-led").className = (busy ? "led-yellow" : (power ? "led-green" : "led-gray"));
		$("atx-hdd-led").className = (hdd ? "led-red" : "led-gray");
	};

	var __updateButtons = function(enabled) {
		for (let id of ["atx-power-button", "atx-power-button-long", "atx-reset-button"]) {
			tools.el.setEnabled($(id), enabled);
		}
	};

	var __clickAtx = function(button) {
		let click_button = function() {
			tools.httpPost("api/atx/click", {"button": button}, function(http) {
				if (http.status === 409) {
					wm.error("Performing another ATX operation for other client.<br>Please try again later.");
				} else if (http.status !== 200) {
					wm.error("Click error", http.responseText);
				}
			});
			__recorder.recordAtxButtonEvent(button);
		};

		if ($("atx-ask-switch").checked) {
			wm.confirm(`
				Are you sure you want to press the <b>${tools.escape(button)}</b> button?<br>
				Warning! This could cause data loss on the server.
			`).then(function(ok) {
				if (ok) {
					click_button();
				}
			});
		} else {
			click_button();
		}
	};

	__init__();
}

File: kvmd/apps/kvmd/api/atx.py
# ========================================================================== #
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2023  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
# ========================================================================== #


from aiohttp.web import Request
from aiohttp.web import Response

from ....htserver import exposed_http
from ....htserver import make_json_response

from ....plugins.atx import BaseAtx

from ....validators.basic import valid_bool
from ....validators.kvm import valid_atx_power_action
from ....validators.kvm import valid_atx_button


# =====
class AtxApi:
    def __init__(self, atx: BaseAtx) -> None:
        self.__atx = atx

    # =====

    @exposed_http("GET", "/atx")
    async def __state_handler(self, _: Request) -> Response:
        return make_json_response(await self.__atx.get_state())

    @exposed_http("POST", "/atx/power")
    async def __power_handler(self, request: Request) -> Response:
        action = valid_atx_power_action(request.query.get("action"))
        wait = valid_bool(request.query.get("wait", False))
        await ({
            "on": self.__atx.power_on,
            "off": self.__atx.power_off,
            "off_hard": self.__atx.power_off_hard,
            "reset_hard": self.__atx.power_reset_hard,
        }[action])(wait)
        return make_json_response()

    @exposed_http("POST", "/atx/click")
    async def __click_handler(self, request: Request) -> Response:
        button = valid_atx_button(request.query.get("button"))
        wait = valid_bool(request.query.get("wait", False))
        await ({
            "power": self.__atx.click_power,
            "power_long": self.__atx.click_power_long,
            "reset": self.__atx.click_reset,
        }[button])(wait)
        return make_json_response()
